### ISSUE TITLE
Portable implementation of let-optionals* broken if the default has side effects

### ISSUE THREAD

**Issue by mnieper** [ashinn/chibi-scheme#778]
The portable implementation of `let-optionals*` shipped with Chibi has been cited as an argument against identifier macros.  The argument is based on the claim without identifier macros, identifiers (in expression context) are just ordinary variable references, and that this knowledge can be used to simplify (some) macro implementations. However, due to the existence of `set!`, even ordinary variable references are not as simple as it seems in Scheme. Thus even without identifier macros, macros, in general, have to bind the values of variable references to temporary variables.

This is not done in the portable `let-optionals*` macro: https://github.com/ashinn/chibi-scheme/blob/master/lib/chibi/optional.sld#L35. This is incorrect because evaluating `default` can modify the value of `tmp`. A simple fix is to bind the value of `tmp` to another temporary variable first.

PS By searching the source tree, I have just noticed that the definition in `init-7.scm` is also affected. (The same seems to be true for the definitions in SRFI 159 and SRFI 166, but there it doesn't matter (unless someone is copying code) because they are not exposed to the world.)

**Comment by ashinn** [comment]
The problem you are concerned about with `default` mutating the input are not resolved by binding a temporary variable.  `default` could just as easily use `set-car!` or `set-cdr!` to mutate the input.  It is thus sufficient to state that it is an error if `default` mutates the input.  This is a common restriction.  I'm not sure if this restriction is necessary, though, given that it will not cause the code to crash.

**Comment by mnieper** [comment]
These two things are separate (and I have to apologize for not stating it precisely in my original post).  While `set!` mutates a binding, `set-car!`  and `set-cdr!`  mutate locations. The two things touch different parts of the specification of `let-optionals*` . Of course, without a specification, everything is allowed, but we can make assumptions about the intended semantics. The syntax (leaving out the rest argument for simplicity) is:

```scheme
(let-optionals* <expression> ((<variable> <init>) ...) <body>)
```

where the tokens have the meanings from R7RS.  The semantics could now state that `<expression>` is evaluated (once) to yield a list and that this list is then destructured.  This is probably the semantics that would be expected.  Alternatively, the semantics could state that `<expression>` may be evaluated more than once.  However, this can hardly be the intended meaning because then, your `let-optionals*` implementation wouldn't (have to) bind an `<expression>` of the form `(<op> . <arg>)` to a temporary.  (Moreover, this would have made the argument against identifier macros based on `let-optionals*` nonsensical.)

So the intended meaning must be assumed to be that `<expression>` is evaluated only once.  But in that case, the macro implementation has to bind even a variable reference to a temporary variable (which, not completely incidentally, also solves the purported identifier macro problem). (Note that it is nonsensical to state something along the lines that it is an error for any (evaluation of) `<init>` to mutate a variable binding occurring in `<expression>`.  Not only would such a constraint alien, but it would also be moot because of `<expression>` being evaluated first (and once).)

What I haven't touched yet is whether any (evaluation of) `<init>` is allowed to mutate the resulting list value (from evaluating `<expression>`) as this is an orthogonal thing.  Either, one could state that it is an error, or the implementation needs to copy the list somehow.  In SRFI 227, I have opted for the latter (but that's a different question not related to the issue I raised here).

**Comment by ashinn** [comment]
Yes, the intended semantics is that `<expression>` is evaluated only once, and this is what is implemented (for systems without identifier syntax).  There is no need to bind it to a temporary variable.

As for the issue of mutation, I think copying the list is overkill.  Leave it unspecified - it would be an extremely bizarre thing for the user to do but presumably they would know what they were doing, and it can be done sensibly.  But the conservative approach of stating that it is an error is fine as well.  Copying the list (which can be arbitrarily long) is a level of paranoia that seems more intended to discourage use of `let-optionals*` than any other goal.

**Comment by mnieper** [comment]
> Yes, the intended semantics is that `<expression>` is evaluated only once, and this is what is implemented (for systems without identifier syntax). There is no need to bind it to a temporary variable.

That's not what is implemented.  In the macro, `tmp` is evaluated more than once. While evaluating a variable reference (which is a form of "identifier syntax", by the way) does not cause side effects, it is affected by side effects. If `tmp` were evaluated only once, the possible occurrence of `set!`  in the dynamic extent of the evaluation of `default` wouldn't be a problem.
 
> As for the issue of mutation, I think copying the list is overkill. Leave it unspecified - it would be an extremely bizarre thing for the user to do but presumably they would know what they were doing, and it can be done sensibly. But the conservative approach of stating that it is an error is fine as well. Copying the list (which can be arbitrarily long) is a level of paranoia that seems more intended to discourage use of `let-optionals*` than any other goal.

I will keep this in mind.  So far, no one has voiced anything about the proposed semantics in SRFI 227. The current semantics (implicit copying of the list) certainly relies on compiler optimizations for `apply` (which also implicitly copies its last argument).

**Comment by ashinn** [comment]
I've added a note that it is an error to mutate the list (the same restriction used on `map` etc.).  Since the case you are concerned about only happens in this error situation, there is no longer any issue, and in a system without identifier syntax it is impossible to tell whether the variable is referenced more than once.

Note your proposal is missing define-opt, which hides the list making these issues moot, while at the same time being more concise and easier to optimize.

**Comment by mnieper** [comment]
> I've added a note that it is an error to mutate the list (the same restriction used on `map` etc.). Since the case you are concerned about only happens in this error situation, there is no longer any issue, and in a system without identifier syntax it is impossible to tell whether the variable is referenced more than once.

The issue I raised has nothing to do with mutating the list.  It is about _rebinding_ a variable, which is a completely different thing:

```scheme
(let ((x '()))
  (let-optionals* x ((a (begin (set! x '(1 2 3)) 4)) (b 5) (c 6))
    (list a b c)))
```

If the expression `x` were evaluated only once, the above would always evaluate to `(4 5 6)`.

(Note that even without identifier macros, the following two expressions are not equivalent: `(let ((tmp x)) (foo tmp tmp (bar)))` and `(foo x x (bar)` , where `foo` and `bar` evaluate to procedures. The moral of the story being that macros must not try to interpret expressions.)

> Note your proposal is missing define-opt, which hides the list making these issues moot, while at the same time being more concise and easier to optimize.

SRFI 227 has `opt-lambda` (and `opt-lambda*`).

Any `define-opt` syntactic sugar is left to a future SRFI (which may conciliate the various `define` extensions we have seen so far).

**Comment by ashinn** [comment]
Yes, I've understood what you're saying from the beginning.  I consider the phrase "mutate ls" to include setting ls.

**Comment by mnieper** [comment]
> Yes, I've understood what you're saying from the beginning. I consider the phrase "mutate ls" to include setting ls.

That hardly makes sense (part of the problem is that, in the documentation, the expression evaluating to a list and the actual list value are not cleanly separated).

I mean, what would this constraint be if you spelled it out explicitly? It would read like the following:

_`<Expression>` is evaluated once but if `<expression>` happens to be a variable reference, it is an error to rebind the referenced variable in the dynamic extent of any `<init>`._

As I said in the beginning unless there is a formal specification, everything is allowed; but the above is certainly so alien to the semantics of the more familiar Scheme special forms that it can hardly be the intended semantics.

If you really mean the above, your implementation is probably correct (in single-threaded environments and under the assumption that the set of primitive expression types has not been extended) but such strange semantics should not be adopted by any standard.

In any case, I don't understand your reluctance against rebinding `tmp` to another temporary.  Compiler beta-reduction will eliminate this overhead (which is a normal thing for many macros) when unnecessary.  This will give your implementation clear semantics (namely that mutating the list obtained from evaluation `<expression>` would be an error) (and, as a side effect, will also resolve the purported problem with identifier macros).

**Comment by ashinn** [comment]
I've more explicitly stated that set! is disallowed as well.

Part of the problem is that we come from two different worlds in terms of specifications.  I follow the R5RS style of trying to make the "spec" (in this case just the (chibi optional) library docs) as minimal as possible, just as I attempt to do with the language itself.  That is, I try not to say any more than need be, omitting anything than follows directly or can reasonably be inferred.

This is an uncommon approach in language specifications, but R5RS is uncommon.  It is admittedly unfriendly to those trying to quickly understand the spec.

Regarding the distinction between mutating objects and assigning to variables, I accept that these are technically different things (and in fact languages such as C++ and Rust allow you to restrict each of them separately on any variable), however I don't consider it "alien to Scheme" to conflate the two.  In the spirit of lambda the ultimate, closures are the basis of all abstraction, and mutating a data structure is akin to setting a closed variable.

The reluctance against introducing a tmp variable is because it's not needed - the notion of an <init> mutating <expression> is so pathological that I would never go out of my way to make it possible.  This keeps the macro simpler.

**Comment by mnieper** [comment]
While I indeed believe that technically separate things should be kept apart and that your new formulation makes hardly sense in this regard (in your formulation, `ls` stands for a value and for a variable at the same time, although both are different kinds of entities), I can infer what you mean.

But when it can be inferred, it can be spelled out as I did above. And then you get the strange constraint I wrote in italics, a constraint like nothing you find anywhere in previous reports (or SRFIs).

> This is an uncommon approach in language specifications, but R5RS is uncommon. It is admittedly unfriendly to those trying to quickly understand the spec.

Can you show me any place where R5RS does a conflation on such a level?

> Regarding the distinction between mutating objects and assigning to variables, I accept that these are technically different things (and in fact languages such as C++ and Rust allow you to restrict each of them separately on any variable), however I don't consider it "alien to Scheme" to conflate the two. In the spirit of lambda the ultimate, closures are the basis of all abstraction, and mutating a data structure is akin to setting a closed variable.

When I used the phrase "alien to Scheme", I meant the constraint itself, not necessarily the wording of it. In a given _implementation_ of Scheme, mutating a data structure and setting a (closed) variable will most likely be done alike (many implementations replace variables that are in the scope of `set!` with boxes), but this does not extend to the semantics of the Scheme programming language. In the Scheme programming language itself (whether specified by R4RS, R5RS, R6RS, or R7RS), rebinding a variable is very different from mutating the store (see also the formal semantics in the appendix of R7RS), so the meaning of "akin" is not at all clear.

> The reluctance against introducing a tmp variable is because it's not needed - the notion of an <init> mutating <expression> is so pathological that I would never go out of my way to make it possible.

I assume you mean "the notion of an `<init>` rebinding `<expression>` if `<expression>` happens to be a variable reference". Now, by the same measure, the following would also be pathological:

```scheme
(let ((a 0))
  (let ((x a))
    (set! a 1)
    x))
```

I don't think you would allow an implementation of `let` that would not cover this "pathological" case.

> This keeps the macro simpler.

We are only talking about inserting `(let ((tmp1 tmp)) <>)`, hardly any complication.

Given this, the real question is what makes the semantics simpler, more coherent, and more in line with the semantics of the rest of the Scheme programming language.  And this is, for sure, the semantics without the extra constraint about rebinding an expression that just happens to be a variable reference. (Just try to write out formal semantics in the style of the appendix of the R7RS for both options.)

**Comment by ashinn** [comment]
I never claimed this particular constraint had precedence, only that R5RS used terse wording and that I personally find this a very natural conflation.  I don't agree with your verbose wording because I prefer a terser style, but if you must spell it out I think the following is better:

"`<Expression>` is evaluated once.  It is an error to call `set!` on `<expression>` or mutate its result in the dynamic extent of any `<init>`."

This is not only simpler but more general.  Usually it's already an error to `set!` a non-variable, but if the Scheme supports SRFI 17 then we can't restrict it to variable references.

There isn't precedence for this restriction because it's only relevant for syntax, and because the only macros in the standards are quite trivial.  That is neither an argument for or against it.

I don't understand what your supposedly pathological example is supposed to show - it's not the same thing.  Show me a use of `let-optionals*` where the list expression is reassigned and we have something to discuss.

**Comment by mnieper** [comment]
> I never claimed this particular constraint had precedence, only that R5RS used terse wording and that I personally find this a very natural conflation.

I find it to be a very dangerous conflation that does not do the Scheme language any favor (and I agree here with Felleisen, Sperber, and Co. on the SRFI 17 mailing list). But we don't have to argue about what wording we like better personally because this is bound to be subjective.

> I don't agree with your verbose wording because I prefer a terser style, but if you must spell it out I think the following is better:
> 
> "`<Expression>` is evaluated once. It is an error to call `set!` on `<expression>` or mutate its result in the dynamic extent of any `<init>`."
> 
> This is not only simpler but more general. Usually it's already an error to `set!` a non-variable, but if the Scheme supports SRFI 17 then we can't restrict it to variable references.

The `set!` exported by the base library and the `set!` of SRFI 17 are very different things (and shouldn't have been given the same name, but that's an aside). They operate on different things and have different effects. The latter does modify a value, so it is already covered by making "mutate its result" an error. (And they are so different that it hardly makes sense to have a single instance of `set!`  covering both meanings in one sentence.)

In any case, the formulation "`<expression>` is evaluated once and it is an error to mutate its result in the dynamic extent of any `<init>`" is definitely shorter.

But what's more important than the length of the wording (whose form is a matter of personal style) is the complexity of the semantics.  Let me say it again: When you write down the formal semantics for both options, the version with the simpler constraint (not mentioning `set!`) is, in fact, simpler by far.

> There isn't precedence for this restriction because it's only relevant for syntax, and because the only macros in the standards are quite trivial. That is neither an argument for or against it.

I don't understand this. Firstly, the `let-optionals*` macro is more trivial (by whatever measure) than some other macros used to implement standard expression types. Secondly, whether the standard expression types are implemented as macros or not seems irrelevant.

> I don't understand what your supposedly pathological example is supposed to show - it's not the same thing.

The same principle underlies it.

> Show me a use of `let-optionals*` where the list expression is reassigned and we have something to discuss.

That's irrelevant for the question of what a good specification is.  A good specification is one where unnecessary complexity is removed and which only depends on a few and simple principles. (And it is certainly Schemy to remove unnecessary restrictions.)

A good specification does not prohibit writing stupid code or even improbable code. An implementation may choose not to support all corner cases of the specification if the cost seems too high and if the corner case is pathological enough, that wouldn't make a difference in practice.

In the case at hand, however, turning `let-optionals*` into an implementation of the simpler semantics has an associated cost of virtually zero.

**Comment by ilammy** [comment]
Excuse me for barging into a flame war, but I think it's important here to distinguish `let-optionals*`-the-Chibi-built-in and `let-optionals*`-as-specified-by-SRFI-227.

If the first one wants to make it undefined behavior to mutate `<expression>` for the sake of more simple implementation, I personally believe that's a valid implementation choice. It's not bound by the semantics prescribed by SRFI 227. There is no spec descended from the heavens that Chibi must follow for its own built-ins. For the one in SRFI 227 – maybe, but that's not the topic of the argument here, right? The topic is whether `let-optionals*` built-in must be changed to follow SRFI 227 semantics to the letter.

Also, it's unlikely there will be a practical difference in actual programs that people would want to write. But *of course* there is a pedantical difference in cherry-picked examples involving the undefined behavior.

**Comment by mnieper** [comment]
Thanks for chiming in but let me put in front that according to the usual understanding of the idiom we are not having a "flame war" here.

You are right in that this discussion is detached from SRFI 227.  Moreover, `let-optionals*` in Chibi is much older than SRFI 227. So the topic is not even about whether `let-optionals*` of Chibi "must be changed to follow SRFI 227 semantics to letter". (And even if it were, the SRFI 227 semantics are not yet set in stone because we are at the beginning of the draft period.)

This is solely about what I would consider an improvement of Alex's Chibi implementation. I have been trying to convince him that more would be gained by simplifying the semantics of his `let-optionals*` than leaving it as is. My point is that the more complicated semantics do not allow a simpler implementation so there is no objective reason for sticking to it. But, of course, the final decision is his. But besides that, it's just an academic discussion where I mean "academic" not in the derogative sense. Scheme is not only a language for practical programming, but also for programming language research and a language for writing down algorithms.  For the latter two cases, clear, simple, and consistent semantics are important.

PS: SRFI 227 is insofar related as writing this SRFI caused me to look at existing `let-optionals*` implementations. And the discussion here may shape the semantics of SRFI 227, of course. Another reason for looking at Chibi's implementation was a claim that having identifier macros would make implementing `let-optionals*`  significantly harder. I would claim that 99% of the macros that don't work without modification in the presence of identifier macros have weird, unorthodox semantics. (The reason for that claim is that macros trying to understand Scheme code are doomed to fail except in the most trivial cases. It's also not a good technique because ignores extensions of the set of expression types, assuming a closed world.) In the case of the `let-optionals*` implementation, two birds would be killed with one stone: The semantics would be simpler and it would be identifier macro-ready.

**Comment by ashinn** [comment]
You don't have to like SRFI 17, but recognize that some implementations support it, and in general the spirit of R7RS is to preserve the ability of people to use Scheme as a vehicle for experimentation.  The standard should allow such extensions, not restrict them with the view that there is only one way to do things.

Even setting SRFI 17 aside, a potentially useful optimization when passed a variable reference is not to introduce any temporary bindings and access the arguments with car, cadr, caddr, etc.  In this case, `set!`ing the variable in one of the `init`s would invalidate the optimization.

Also, although it is straightforward to modify the current `let-optionals*` to introduce a temp binding, in the current implementation it would require a helper macro, and when it comes to the Chibi core I count lines and try very hard to keep things to a minimum.  Since the only case where the distinction being discussed matters is so bizarre I would actually prefer to deliberately make the code break, I don't have much motivation to change this.  [Basically I consider changing the Chibi core out of the question at this point, and consider the topic to be the semantics for SRFI 227.]

Note we can also consider similar restrictions in hypothetical future macros.  One case is `match`.  `set!`ing the variable matched can similarly cause problems, and this is specifically a case where you can't wrap the input in a temporary binding because the `set!` pattern can be used to refer to and later modify (after pattern matching has completed) the input variable reference.  [The actual final semantics of SRFI 204 may not have this issue, but that's beside the point.]

So restrictions on `set!`, with or without mutation restrictions, are something we should probably become more comfortable with.

**Comment by mnieper** [comment]
> You don't have to like SRFI 17, but recognize that some implementations support it, and in general the spirit of R7RS is to preserve the ability of people to use Scheme as a vehicle for experimentation. The standard should allow such extensions, not restrict them with the view that there is only one way to do things.

I do think that SRFI 17 is superfluous and that it should have at least chosen a different keyword than `set!`. But my personal opinion on SRFI 17 is irrelevant here, I think. I don't see where I said something against using Scheme for language experimentation or about forbidding such extensions. (As I said above, a good implementation does not prohibit doing "stupid" things.) Anyway, SRFI 17 does not have anything to do with the `let-optionals*` issue raised in this thread, so I think we can drop this particular part of the discussion as it does not belong here.

> Even setting SRFI 17 aside, a potentially useful optimization when passed a variable reference is not to introduce any temporary bindings and access the arguments with car, cadr, caddr, etc. In this case, `set!`ing the variable in one of the `init`s would invalidate the optimization.

Such an optimization should not be the task of the macro but of the actual compiler. Beta reduction is probably the first optimization implemented in any Scheme compiler/interpreter. In fact, a lot of standard macros rely on this optimization. 

See O. Waddell's PhD (https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.71.363&rep=rep1&type=pdf) that points out why extended syntactic abstraction increases the need of certain compiler optimizations. (This is related to the fact, that code-walking macros that do the optimization themselves are not really supported by the Scheme expansion model.)

> Also, although it is straightforward to modify the current `let-optionals*` to introduce a temp binding, in the current implementation it would require a helper macro, and when it comes to the Chibi core I count lines and try very hard to keep things to a minimum. 

If you rely on beta reduction, you don't need a helper macro. And if not, you would just need one more clause with an (otherwise inaccessible) pattern literal. And if you would really count line numbers, you would probably first try to remove the many copies of the `let-optionals*` macro.

> Since the only case where the distinction being discussed matters is so bizarre I would actually prefer to deliberately make the code break, I don't have much motivation to change this. [Basically I consider changing the Chibi core out of the question at this point, and consider the topic to be the semantics for SRFI 227.]

If it is out of the question for you anyway, it probably makes sense for me to withdraw from this discussion here because it would lead nowhere. SRFI 227 semantics should be discussed on the SRFI 227 mailing list.

> Note we can also consider similar restrictions in hypothetical future macros. One case is `match`. `set!`ing the variable matched can similarly cause problems, and this is specifically a case where you can't wrap the input in a temporary binding because the `set!` pattern can be used to refer to and later modify (after pattern matching has completed) the input variable reference. [The actual final semantics of SRFI 204 may not have this issue, but that's beside the point.]

That semantics would be so weird, that I would not even consider it. Variables are not boxes. If you need a boxed value, use a box.

Besides, such a match form would be incompatible with identifier macros, but that would not be the fault of the identifier macros but the fault of the weird semantics of such a match form (which seems tailored to foil this or any other extension of the set of expression types).

> So restrictions on `set!`, with or without mutation restrictions, are something we should probably become more comfortable with.

If we want Scheme to become a second PHP, then yes.  If not, I'd like to keep the clear semantics of Scheme.

**Comment by ashinn** [comment]
I've already stated that I consider all mutation related, whether via `set!` or boxes, both at the implementation level and at the user level.  I'm also of the opinion that all mutation should be minimized, the reasons for this being many.  Distinguishing between types of mutation and picking and choosing which are allowed does nothing to help the implementation or the understanding of a program, and merely confuses the user who must keep track of which is allowed.

But I'm repeating myself now, and you've made an unjust and uncalled for insult by comparing my views to PHP.  There can be no more civil discussion at this point.

**Comment by mnieper** [comment]
> I've already stated that I consider all mutation related, whether via `set!` or boxes, both at the implementation level and at the user level. 

> I'm also of the opinion that all mutation should be minimized, the reasons for this being many.

I fully agree. In fact, I can think of a lot of reasons why `set!` should be deprecated in future Scheme standards.

> Distinguishing between types of mutation and picking and choosing which are allowed does nothing to help the implementation or the understanding of a program, and merely confuses the user who must keep track of which is allowed.

One is mutating the environment, which is a semantic concept, which is not directly accessible, the other is mutating the store, which is reified through the locations allocated by the objects being mutated.

Any user of Scheme has to understand the difference, and you are not doing those users who have difficulties with the two concepts a favor by conflating the two things.

```scheme
(define (f! x) (set! x 1))
(define x 0)
(f! x)
x ;=> 0

(define (g! x) (set-box! x 1))
(define x (box 0))
(g! x)
(unbox x) ;=> 1
````

> But I'm repeating myself now, and you've made an unjust and uncalled for insult by comparing my views to PHP. There can be no more civil discussion at this point.

My PHP comparison (that was clearly not about your views (it would have been a comparison across categories) but of this particularly IMO weird semantics) was obviously exaggerated to make a point. In any case, how can the fact that I perceive the "weird" semantics for `match` as something that would befit PHP be a personal insult toward you?`

Scheme has formal semantics for some reason.  If you add things that make the formal semantics much more complicated, it is probably the wrong thing to add.

**Comment by ashinn** [comment]
*sigh*  There never was a bug in Chibi to begin with.  I really need to insist on test cases, and reject any issue that doesn't include one.

In the portable R7RS impl there was a potential interaction depending on the order of evaluation of lambda arguments.  I still disagree that this is a bug - you can't cause a crash, and anyone mutating the list they are processing presumably knows what they are doing.  However, since forcing the order is trivial, involving not a single new line of code, and is outside of the Chibi core, I've gone ahead and made the change - with a test case!
